app/scripts/migrations/144.ts

```diff
@@ -0,0 +1,301 @@
+      ),
+    );
+
+    return state;
+  }
+
+  if (!permissionControllerState || !isObject(permissionControllerState)) {
+    global.sentry?.captureException?.(
+      new Error(
+        'Skipping migration: `PermissionController` state not found or is not an object.',
+      ),
+    );
+
+    return state;
+  }
+
+  if (!isObject(permissionControllerState.subjects)) {
+    global.sentry?.captureException?.(
+      new Error(
+        'Skipping migration: `PermissionController.subjects` state is not an object.',
+      ),
+    );
+
+    return state;
+  }
+
+  if (
+    !selectedNetworkControllerState ||
+    !isObject(selectedNetworkControllerState)
+  ) {
+    global.sentry?.captureException?.(
+      new Error(
+        'Skipping migration: `SelectedNetworkController` state not found or is not an object.',
+      ),
+    );
+
+    return state;
+  }
+
+  if (!isObject(selectedNetworkControllerState.domains)) {
+    global.sentry?.captureException?.(
+      new Error(
+        'Skipping migration: `SelectedNetworkController.domains` state is not an object.',
+      ),
+    );
+
+    return state;
+  }
+
+  const currentChainId = getChainId(
+    networkControllerState.selectedNetworkClientId,
+    networkControllerState.networkConfigurationsByChainId,
+  );
+
+  if (typeof currentChainId !== 'string') {
+    global.sentry?.captureException?.(
+      new Error(
+        'Skipping migration: The chain ID resolved from the `NetworkController.selectedNetworkClientId` value is not a string.',
+      ),
+    );
+
+    return state;
+  }
+
+  const updatedSubjects: string[] = [];
+
+  // Add permission to use the current globally selected network to all Snaps
+  // that have the `endowment:ethereum-provider` permission.
+  const entries = Object.entries(permissionControllerState.subjects) as [
+    string,
+    GenericPermissionControllerSubject,
+  ][];
+
+  permissionControllerState.subjects = entries.reduce<
+    PermissionControllerSubjects<PermissionConstraint>
+  >((accumulator, [key, subject]) => {
+    const permissionKeys = Object.keys(subject.permissions);
+    const needsMigration = permissionKeys.includes(
+      SnapEndowments.EthereumProvider,
+    );
+
+    if (!needsMigration) {
+      return {
+        ...accumulator,
+        [key]: subject,
+      };
+    }
+
+    updatedSubjects.push(key);
+
+    const caveatValue = addPermittedEthChainId(
+      getCaveat(subject),
+      currentChainId as `0x${string}`,
+    );
+
+    const newSubject: GenericPermissionControllerSubject = {
+      ...subject,
+      permissions: {
+        ...subject.permissions,
+        [Caip25EndowmentPermissionName]: {
+          caveats: [
+            {
+              type: Caip25CaveatType,
+              value: caveatValue,
+            },
+          ],
+          date: Date.now(),
+          id: nanoid(),
+          invoker: key,
+          parentCapability: Caip25EndowmentPermissionName,
```

Comment:
probably doesn't matter but should we inherit this metadata from the existing CAIP-25 permission if it exists?